tokio/src/sync/mod.rs

```diff
@@ -431,6 +431,24 @@
 //!   number of permits, which tasks may request in order to enter a critical
 //!   section. Semaphores are useful for implementing limiting or bounding of
 //!   any kind.
+//!
+//! # Runtime compatibility
+//!
+//! All synchronization primitives provided in this module are runtime agnostic.
+//! You can freely move them between different instances of the Tokio runtime
+//! or even use them from non-Tokio runtimes.
+//!
+//! When used in a Tokio runtime, the synchronization primitives participate in
+//! [cooperative scheduling](crate::task#cooperative-scheduling),
+//! periodically yielding back control to the runtime to avoid starving it.
+//! In a non-Tokio runtime, the yield points where control is handed back do
+//! nothing, not changing the behavior of the synchronization primitive in any
+//! way.
+//!
+//! The only exception to runtime compatibility is the
+//! [`mpsc::Sender::send_timeout`] method, which can't be called from non-Tokio
+//! runtimes. However, [`Senders`](mpsc::Sender) can still be moved between
+//! different instances of the Tokio runtime.
```

Comment:
I think it would be an improvement to not talk about `send_timeout` specifically. That way, we do not have to change the wording here if new `_timeout` methods are added in the future.
```suggestion
//! As an exception, methods ending in `_timeout` are not runtime agnostic
//! because they require access to the Tokio timer. See the documentation of
//! the `*_timeout` method for more information on its use.
```

Reply:
I like it. I was wondering how we could avoid extra churn if there are more exceptions added in the future. My initial idea was to add a warning message like [this](https://github.com/tokio-rs/tokio/blob/91169992b2ed0cf8844dbaa0b4024f9db588a629/tokio/src/sync/mpsc/mod.rs#L83-L84) that says there might be more exceptions added in the future, but it felt too over the top.

Reply:
You think it would be a good idea to at least refer to `mpsc::Sender::send_timeout` as an example?

```suggestion
//! As an exception, methods ending in `_timeout` (like
//! [`mpsc::Sender::send_timeout`]) are not runtime agnostic because they
//! require access to the Tokio timer. See the documentation of each
//! `*_timeout` method for more information on its use.
```

Reply:
I suppose we could ... but I don't think it's necessary. I also try to avoid parentheticals if possible.